Integrating codemirror 6

davidcarlisle opened the pull request:
<strike>This PR is marked as draft but</strike> I believe it's already usable

@niruvt  @iamalinguist I guess this would mostly affect your languages.
 
A rendered version of the site showing the new code is at

https://davidcarlisle.github.io/learnlatex.github.io

## Changes

* any page can have `editor: cm6` or `editor: ace` in the yaml block at the top. The default is `ace` unless `lang` is `mr` or `hi`.
* A custom codemirror 6 bundle has been constructed that includes the "stex" tex lexer to enable tex syntax highlighting. (This javascript is checked in, it may be better to serve separately from eg texlive.net, we'll see....
* A modified javascript file, `runlatex-cm6.js` is loaded that loads codemirror. (It would be able to combine this with `runlatex.js` and have conditional editor code, but simpler for experiments to have them separate for now.)
* Overleaf and texlive.net upload works, including the multi-file lesson-13
* Line selection works in lesson 15
* Automatic preamble insertion works (not used at learnlatex)
* One light and one dark theme are implemented, choosable in the settings page (for en and mr only so far)
* runlatex.editorlines is interpreted as a multiple of 1.4em and used to set the editor height.

## For testing

* The Marathi pages default to code mirror as noted, <strike>except for mr/lesson-3 which forces ace and en/lesson-03 which forces codemirror (just to test in-page `editor: cm6` settings work)</strike>
* <strike>All pages that load the codemirror version get a red border just for ease of checking.</strike>

## Not done yet

* <strike>Editor themes (I think a dark theme is included in the bundle constructed but it's not accessed yet)</strike>
*  Editor mode. Currently it is always in tex highlighting mode, some examples using ace turn that off and use a plain text mode
* command completion, ace has some support for command completion
* <strike>Editor size: the runlatex editor lines configuration option isn't used</strike>
* Optimising page loading: currently switching in the editor causes some page flicker as the original text is swapped out and the editor added.
* Editor scripting:  <strike>Lesson 15 uses the ACE API to select certain lines in the display</strike> and the Help page uses the ace API to show the list of available options. Neither have any codemirror version setup yet.
* Checking status of `<tab>` for accessibility use (the ace setup avoids stealing tab keybindings so it remains available for accessibility focus control)

niruvt commented:
Hello @davidcarlisle.

Thanks a lot for all the work.

I have some questions.

1. > any page can have editor: cm6 or editor: ace in the yaml block at the top. The default is ace unless lang is `mr` or `hi`.

    * What's the reason for preferring ace over cm6 as default for other languages? (just curious)
    * Most languages with complex ligatures, i.e., all of Indic languages (and as you must be aware, we have _plenty_ of them :P) but also, e.g., Arabic (try `كُلُّ نَفْسٍۢ ذَآئِقَةُ ٱلْمَوْتِ ۗ وَنَبْلُوكُم بِٱلشَّرِّ وَٱلْخَيْرِ فِتْنَةًۭ ۖ وَإِلَيْنَا تُرْجَعُونَ` on the main site).
    * This effectively means, we don't need to mark `hi` and `mr` as special as there are many such languages. This also means that the default editor will lack support for all such languages, so unless there are really strong reasons in support of it, probably we should consider making cm6 the default.
2. > Editor themes (I think a dark theme is included in the bundle constructed but it's not accessed yet)

    * Do we otherwise have dark theme support as of now?

davidcarlisle commented:
@niruvt 
> 
>     1. > any page can have editor: cm6 or editor: ace in the yaml block at the top. The default is ace unless lang is `mr` or `hi`.
>  
>        * What's the reason for preferring ace over cm6 as default for other languages? (just curious)

I wanted to be cautious while testing, I suspect that in the end I will switch the defaults or just offer code mirror.

Currently for example the "automatic preamble" code (which isn't used at learnlatex but is elsewhere) doesn't work eg

https://texlive.net/run?hello

you can just type `hello` and a default latex document preamble gets filled in. The additions are added via the ACE API, I'm sure codemirror can do the same, but I need to work through it.
Also I think the tex support is probably going to be better in ace (unless someone writes a new tex parser module for codemirror) it seems everyone uses the "stex" module I have used here which is OK but has less support than the ace one.
Given that ace can't be used at all for some languages I think that's OK, but I can't just "silently" switch all existing languages.

>        * Most languages with complex ligatures, i.e., all of Indic languages (and as you must be aware, we have _plenty_ of them :P) but also, e.g., Arabic (try `كُلُّ نَفْسٍۢ ذَآئِقَةُ ٱلْمَوْتِ ۗ وَنَبْلُوكُم بِٱلشَّرِّ وَٱلْخَيْرِ فِتْنَةًۭ ۖ وَإِلَيْنَا تُرْجَعُونَ` on the main site).
>        * This effectively means, we don't need to mark `hi` and `mr` as special as there are many such languages. This also means that the default editor will lack support for all such languages, so unless there are really strong reasons in support of it, probably we should consider making cm6 the default.

Yes I suspect this will be the end result. But note that `lang` in that test isn't a test of an arbitrary user-settable language it is a setting of the yaml variable set in the page, so the only possible values are the values listed in the learnlatex language switcher. If someone adds arabic then that language code would need to be added in lots of places, so adding it here as well wouldn't be so bad. (but I may just switch anyway as I say, it depends a bit on how many features I manage to recreate in codemirror)

> 
>     2. > Editor themes (I think a dark theme is included in the bundle constructed but it's not accessed yet)
>              
>        * Do we otherwise have dark theme support as of now?

maybe later today. I thought I understood the documentation enough to include one in the bundle I packaged and to access it, but apparently not:-) It seems though I am slowly starting to understand the cm6 setup and things fall in to place, so I hope to 
have the theme option this weekend at the latest.

For ace there are multiple themes that you can access at 
https://www.learnlatex.org/en/settings
as far as I understand, for codemirror I get to choose the list of themes I want to bundle into the cm6 version I distribute then can offer that choice on the site, I don't think I'll offer many but I wanted to start with at least a choice of a light theme and a dark theme.

niruvt commented:
> I wanted to be cautious while testing, I suspect that in the end I will switch the defaults or just offer code mirror.

Makes sense.

> Currently for example the "automatic preamble" code (which isn't used at learnlatex but is elsewhere) doesn't work eg
> 
> [...]
> 
>  Also I think the tex support is probably going to be better in ace (unless someone writes a new tex parser module for codemirror) it seems everyone uses the "stex" module I have used here which is OK but has less support than the ace one.

Ahh, okay okay. Thanks for pointing these out.

> Yes I suspect this will be the end result. But note that lang in that test isn't a test of an arbitrary user-settable language it is a setting of the yaml variable set in the page, so the only possible values are the values listed in the learnlatex language switcher. If someone adds arabic then that language code would need to be added in lots of places, so adding it here as well wouldn't be so bad. (but I may just switch anyway as I say, it depends a bit on how many features I manage to recreate in codemirror)

Understood. Makes sense.

> as far as I understand, for codemirror I get to choose the list of themes I want to bundle into the cm6 version I distribute then can offer that choice on the site, I don't think I'll offer many but I wanted to start with at least a choice of a light theme and a dark theme.

Great!

Thanks.

davidcarlisle commented:
@niruvt 

![image](https://github.com/user-attachments/assets/00af13a3-d81b-4dc8-97e8-f12ae5136f63)


Dark mode...

davidcarlisle commented:
I think this can be merged.

* apart from some very small changes in `runlatex.js` encapsulating the ACE API in runlatex-specific functions, this has no effect on most pages. It simply adds another variant javascript, `runlatex-cm6.js` and some setup options to select it.
* I lesson 15 I simplified the selection to just highlight the error line, not move the cursor to the column location (as it simplifies the code and never seemed that useful)
* All existing languages keep ace, except Marathi.
* Currently the codemirror version does not supply folding for latex environments and does not offer command completion. Codemirror6 has both those features but it wasn't clear to me if they can be enabled on the "stex" module being used which is a legacy codemirror5 lexer. These features could be added later.
* The codemirror version currently uses the tex syntax highlighter in all cases. With ACE, I switched to plain text mode for log files and other examples without a tex run interaction. This must also be possible to add later, but isn't critical for this PR.
* I removed the cm6 bundle file as it's a large generated blob that's not good for git and would need duplicating here and in the latexcgi repo and elsewhere. The files now reference a copy at texlive.net. (We never used a local copy of ACE., always referencing a copy from cdnjs.cloudflare.com.)

niruvt commented:
@davidcarlisle 

It requires translation, doesn't it? I will add it tomorrow. Thanks.

niruvt commented:
कोडमिरर-६ ह्या मजकूर-संपादकासाठीची रंगछटा (गडद/उजळ).

davidcarlisle commented:
> कोडमिरर-६ ह्या मजकूर-संपादकासाठीची रंगछटा (गडद/उजळ).

@niruvt  thanks: added.

davidcarlisle commented:
@niruvt  sorry this took so long...

It's merged now and codemirror should be live,  you may want to change the text below.
In most other languages I made a minimal change to make a bullet list of both editors without adding any new text but I didn't dare here:-)


ही उदाहरणे परिपूर्ण आहेत, परंतु तुम्ही त्यांच्यात बदलदेखील करू शकता.  येथे वापरला जाणारा
लाटेक्-संपादक [एस](https://ace.c9.io/) हा आहे.

at https://www.learnlatex.org/mr/help#%E0%A4%89%E0%A4%A6%E0%A4%BE%E0%A4%B9%E0%A4%B0%E0%A4%A3%E0%A5%87-%E0%A4%9A%E0%A4%BE%E0%A4%B2%E0%A4%B5%E0%A4%A3%E0%A5%87


If you are not currently set up to make a PR to this repo you could just drop the markdown replacement for this into this discussion and I'll add. 

```

ही उदाहरणे परिपूर्ण आहेत, परंतु तुम्ही त्यांच्यात बदलदेखील करू शकता.  येथे वापरला जाणारा
लाटेक्-संपादक [एस](https://ace.c9.io/) हा आहे.

```

niruvt commented:
@davidcarlisle yeah, I would just drop the replacement text here, kindly make the change:

```
ही उदाहरणे परिपूर्ण आहेत, परंतु तुम्ही त्यांच्यात बदलदेखील करू शकता.  येथे वापरला जाणारा
लाटेक्-संपादक [कोडमिरर]() हा आहे.
```

What should the link be?? I am not sure. You may add it directly.

No problem with the delay. Thanks for completing it :)

davidcarlisle commented:
done:-)
